### ISSUE TITLE
prioritized channel data processing for limited bandwidth scenario

### ISSUE THREAD

**Issue by nero-ruijie** [neutrinolabs/xrdp#2837]
### xrdp version

0.9.20

### Detailed xrdp version, build options

~~~
ANY
~~~

### Operating system & version

any

### Installation method

Doesn't matter

### Which backend do you use?

any

### What desktop environment do you use?

any

### Steps to reproduce

As a fact, almost all network io events are handled in the single working thread/loop **xrdp_process_main_loop** in xrdp server process and result in any read or write blocking will cause all channels been blocked. 

### ✔️ Expected Behavior

Image channel should has lower priority to sound(send), mouse(recv), etc.  Images can be split into fragments as descried in "3.2.5.9.3.1 Processing Fast-Path Update Fragments" of MS-RDPBCGR - if not misunderstood. Sending images in a passive manner(using sending qeueu or somewhat) may be better.

### ❌ Actual Behavior

It's hard to drag window and sound is broken with bandwidth limited.

**Comment by matt335672** [comment]
Hi @nero-ruijie 

I'm finding it difficult to see how implementing queuing priorities will help here.

You've got a both graphic and sound data coming down from the server to the client. Both of these are critical, and a delay to either will be noticeable by the client. Your example illustrates this well:-

> It's hard to drag window and sound is broken with bandwidth limited.

The window dragging is graphical data and it's coming down the same channel as the sound. If they don't fit in the channel, one or both will necessarily suffer.

xrdp works at a pixel level and has no concept of windows as such. It can't know whether you're dragging a window or watching a video or waiting for a webpage to load. It's just graphical data.

**Comment by nero-ruijie** [comment]
> Hi @nero-ruijie
> 
> I'm finding it difficult to see how implementing queuing priorities will help here.
> 
> You've got a both graphic and sound data coming down from the server to the client. Both of these are critical, and a delay to either will be noticeable by the client. Your example illustrates this well:-
> 
> > It's hard to drag window and sound is broken with bandwidth limited.
> 
> The window dragging is graphical data and it's coming down the same channel as the sound. If they don't fit in the channel, one or both will necessarily suffer.
> 
> xrdp works at a pixel level and has no concept of windows as such. It can't know whether you're dragging a window or watching a video or waiting for a webpage to load. It's just graphical data.

Hi @matt335672 
we just take a look at the following condition:
> Limit bandwidth to 100KBps
> H.264/RFX 50Mbps/33fps stream -> ~190KB per frame (average) -> blocking for 2s

If xrdp sends image data in lower priority, xrdp will be less sensitive to bandwidth limit or giant image blocks - it gets more chance for sending sound output.  The max size of TS_FP_UPDATE_PDU is limited to 16,383 bytes, there must be some reason.

The other problem is why sending data blocks recving or other io instance handling. It's better to make sockets working in real async/non-blocking pattern.

**Comment by matt335672** [comment]
I still can't see that.

In your example, once the frame is processed and the image data is queued, it still has to be sent, whether on not the sound gets sent first. If the sound is prioritised, you'll get better sound, but you'll see more jerkiness in the frame. You have to pick one or the other.

**Comment by nero-ruijie** [comment]
Yes, with prioritised/non-block mode,  we will got better input response, better sound and maybe lower FPS or lower bitrate(with bandwidth detection and dynamic bitrate control) for images. Personally, it's more acceptable for FPS loss than sound broken.

**Comment by matt335672** [comment]
I don't think non-blocking will help much in this particular scenario (although it may in others).

The blocking will happen if xrdp is unable to write to the TCP channel as it's full. At the point we can't do any useful data transfer anyway, so blocking is an entirely reasonable thing to do(?) The codec works in a separate thread so that will continue to do what it can. We could maybe process input events, but as the output pipe is blocked there's no way to feed the effects of these back to the user anyway! Am I missing anything?

I can see that users may have a personal preference for prioritising sound over image data, or maybe vice-versa.

**Comment by nero-ruijie** [comment]
> The blocking will happen if xrdp is unable to write to the TCP channel as it's full. At the point we can't do any useful data transfer anyway, so blocking is an entirely reasonable thing to do(?) 

We can recv drive/fs redirection data while blocking in sending images. So, I think it's unreasonable.

> The codec works in a separate thread so that will continue to do what it can.
> We could maybe process input events, but as the output pipe is blocked there's no way to feed the effects of these back to the user anyway!

If image frame blocks 2s, all mouse events in this 2s (not strictly to say) are handled in a very short time 2s later. Separate two single click could result in double click.  So, we will get better, smoother mouse experience anyway with non-blocking.

**Comment by matt335672** [comment]
Sorry - I've given you the wrong impression. Checking the code, we're not blocked on input when we're blocked on output - these are separate `poll()` events.  

When an output fd is blocked we do one of two things:-
1) Wait a short while and then fail if the message can't wait, or, 
2) Queue the message and wait for the fd to become writeable again. That check happens in parallel with all the other fd's in the process.

So we can process incoming redirection data and mount events to a point. What we can't do is respond to them.

**Comment by nero-ruijie** [comment]
Thanks for your patient clarification. 

> Queue the message and wait for the fd to become writeable again. That check happens in parallel with all the other fd's in the process.

I have read codes of **trans_check_wait_objs**, **trans_write_copy_s** and got how xrdp queued the sending message when blocked and handing other events in parallel. But I also noticed that xrdp looks like using an infinite(equivelant) buffer to queue the pending messages if there are no flow control for the entire client-server data loop (e.g. pulse audio,  looks like there are no policy to handle SNDWAV_CONFIRM PDU in **sound_process_wave_confirm**, is that right?).

**Comment by nero-ruijie** [comment]
Thanks for your clarification.

**Comment by matt335672** [comment]
Sorry for the confusion I caused.

I was working on a fix for a FreeRDP problem yesterday in between chatting to you, and I didn't address either at all satisfactorily.

The code is quite difficult to follow in places. I notice you were interested in flow control earlier. There are mechanisms around this, but again they're quite tricky to piece together. Let me know if you need more information and I'll explain what I can - hopefully better than yesterday.
